TangoShutter: refactor (changes from https://github.com/mxcube/mxcubecore/pull/847/)

mockoocy opened the pull request:
Adds code changes from PR https://github.com/mxcube/mxcubecore/pull/847
And a test case for the TangoShutter Hardware Object.

At MAXIV we use these changes for quite a long time, but it has not been merged in the original for some reason (which is rather unclear looking at the PR's discussion).
The difference between the version at https://github.com/mxcube/mxcubecore/pull/847 and this PR is addition of type annotations and adoption of new linting rules.

mockoocy committed:
Refactor of TangoShutter

Originally added in github.com/mxcube/mxcubecore/pull/847

mockoocy committed:
Adapts TangoShutter to new linting rules and adds type hints

elmjag commented on test/pytest/test_hwo_isara_maint.py:
```diff
@@ -90,7 +90,7 @@ def disable_remote_mode(self):
 def _disconnect_channels(maint: ISARAMaint):
     """
-    disconnect signal callbacks from tango attribute pollers
+    disconnect signal callbacks from tango attribute pollersf
```
What up with this change?

mockoocy replied:
Tragedy in few parts:
1. I opened this file 
2. I wanted to do `ctrl + f` to search for something.
3. I must have missed the `ctrl` and pressed `fn` :D
4. `git add test/`

mockoocy replied:
Removed the change now.

mockoocy committed:
Adds test case for TangoShutter

The ruff config had to be adjusted to ignore the rule saying that
methods have to start with lowercase letter - as there is a mocked
Tango Device.

elmjag commented on test/pytest/test_hwo_tango_shutter.py:
```diff
@@ -0,0 +1,123 @@
+        self._is_open = False
+        super().__init__(*args, **kwargs)
+
+    @command()
+    def Open(self):
+        self._is_open = True
+
+    @command()
+    def Close(self):
+        self._is_open = False
```
Instead of disabling N802 check for this whole file in [ruff.toml](https://github.com/mxcube/mxcubecore/pull/1158/files#diff-26f436cb29eecc7bb4f9066747b43df8da81ac808955ce95c109f5d7aa778989),
disable it only for these two special case methods, like this:

```suggestion
    @command()
    def Open(self):  # noqa: N802
        self._is_open = True

    @command()
    def Close(self):  # noqa: N802
        self._is_open = False
```

fabcor-maxiv replied:
Yes, I was about to comment on this as well.

Otherwise:

```suggestion
    @command()
    def open(self):
        self._is_open = True

    @command()
    def close(self):
        self._is_open = False
```

The upper case "convention" for Tango commands does not really matter here in a test case, I guess. Then you would need to change in the other places in the test case of course.

mockoocy replied:
Done

beteva commented on mxcubecore/HardwareObjects/TangoShutter.py:
```diff
@@ -86,26 +98,18 @@ def init(self):
-            self.config_values = None
-
-    def _update_value(self, value):
+    def _update_value(self, value: DevState) -> None:
```
Are you sure the value is DevState type. I believe this is the output of the DevState command, which is a str.

beteva replied:
Nice with the test.

mockoocy replied:
When I checked the type of value there during runtime I got a `DevState` type. 
I believe that the `state_channel` could be set to return a string (with `TangoChannel.read_as_str` set to `True`).
 I think setting it as `DevState | str` is a safe choice then :)

beteva commented:
Nice with the test

beteva committed:
Merge branch 'develop' into pm-add-tango-shutter-changes

fabcor-maxiv commented:
I recommend you check how the docstrings are rendered:

https://mxcubecore--1158.org.readthedocs.build/en/1158/dev/autosummary/mxcubecore.HardwareObjects.TangoShutter.html#module-mxcubecore.HardwareObjects.TangoShutter

marcus-oscarsson commented:
Oh yes, the documentation looks a bit strange.

marcus-oscarsson commented:
Thanks, alot. I think its really worth taking a look at how the documentation is rendered, there seems to be a tag missing or something similar.

beteva committed:
Update TangoShutter.py

Added code-block to improve documentation.

mockoocy committed:
TangoShutter: fix docstrings

Also changed type hint for `TangoShutter._update_value`.
The value could possibly be a string (i.e. using `TangoChannel` with
`read_as_str` set to True).

mockoocy commented:
Thanks @beteva for fixing the xml code rendering in the docs! :)

I also pushed some changes, information about return rendered wrongly, when there was no space between docstring summary and `Args:, Returns:`, etc. blocks. (`:returns` text appeared in the summary)

fabcor-maxiv commented:
Looks like there is still some issue on the module docstring here: 
![image](https://github.com/user-attachments/assets/2f3fa71a-3478-4e3e-9137-e1be37009c38)

fabcor-maxiv commented on mxcubecore/HardwareObjects/TangoShutter.py:
```diff
@@ -25,30 +25,44 @@
+    - make sure only double quotes are used inside the values dictionary. No single quotes (') are allowed !
+    - the second element of the list should be a standard HardwareObjectState name
+    (UNKNOWN, WARNING, BUSY, READY, FAULT, OFF - see in BaseHardwareObjects.py)!
+The <values> property is optional.
```
```suggestion
list. The new state is added to both the ``VALUES`` and the ``SPECIFIC_STATES`` enumeration.

Attention:

- do not use tuples or the Python JSON parser will fail!
- make sure only double quotes (``"``) are used inside the values dictionary,
  no single quotes (``'``) are allowed!
- the second element of the list should be a standard ``HardwareObjectState`` name
  (``UNKNOWN``, ``WARNING``, ``BUSY``, ``READY``, ``FAULT``, or ``OFF``;
  see in ``BaseHardwareObjects``)!
    
The ``<values>`` property is optional.
```

mockoocy replied:
It works! Thanks :)
Adjusted the linting and commited.
Btw, in the future some ruff rules could be implemented for formatting these docstrings. It seems pretty easy to break these.

fabcor-maxiv replied:
In principle there are (at least) two blocks or Ruff rules for docstring:

* https://docs.astral.sh/ruff/rules/#pydoclint-doc
* https://docs.astral.sh/ruff/rules/#pydocstyle-d

I am not sure if there are/were new violations to these rules here in this pull request. It feels to me like there are/were some violations, and I am surprised Ruff did not complain. I do not know how far Ruff can check that the docstrings are valid in terms of:
* reStructuredText
* Sphinx doc
* Google style

mockoocy committed:
Fix module docstring formatting

Co-authored-by: fabcor <fabien.coronis@maxiv.lu.se>

fabcor-maxiv commented on ruff.toml:
```diff
@@ -7118,3 +7118,6 @@ convention = "google"
     "PT022",
     "S108",
 ]
+"test/pytest/test_hwo_tango_shutter.py" = [
+    "N802"
+]
```
Please, do not add anything in the Ruff ignores. We have to actually fix the rule violation. If it truly is not possible to fix the rule violation, then we can set an inline ignore directly in the Python file.

```suggestion
```

mockoocy committed:
Removes test file from ruff.toml

The exceptions for capital case functions (Tango commands) are moved to
the functions that offend them instead of creating an exception for the
whole file.

fabcor-maxiv commented on mxcubecore/HardwareObjects/TangoShutter.py:
```diff
@@ -114,36 +127,43 @@ def _initialise_values(self):
+            _state = self.get_value().name
+            return self.SPECIFIC_STATES[_state].value[0]
         except (AttributeError, KeyError):
+            logging.exception("Exception in get_state()")
```
This is strange to me.
Feels like it should have triggered Ruff rule `LOG015`: <https://docs.astral.sh/ruff/rules/root-logger-call/>.

```suggestion
            hwr_logger.exception("Exception in get_state()")
```

With somewhere towards the top, after the imports, something like this:

```
hwr_logger = logging.getLogger("HWR")
```

fabcor-maxiv replied:
> Feels like it should have triggered Ruff rule `LOG015`: https://docs.astral.sh/ruff/rules/root-logger-call/.

Ah okay, it says right there that it is a "preview" rule that needs the `--preview` flag to be enabled.

Still should be fixed, though.

mockoocy replied:
Well spotted :D Fixed

mockoocy committed:
TangoShutter: Now uses HWR logger
